### PR TITLE
Run allocation first after BMI stop

### PR DESCRIPTION
**Pull request by visr**
This allows allocation over period `(t, t + dt)` to use variables set over BMI at time `t`.

We no longer run allocation as a callback, but call it ourselves, such that we can control that BMI runs allocation first before running the physical layer.

### CODE REVIEW

**Comment by visr** [comment]
Some pretty large allocation test differences suggest we don't call allocation correctly yet.

```
No Captured Logs for test item "Allocation level control" at test\allocation_test.jl:308 on worker 26268
Error in testset "Allocation level control" on worker 26268:
Test Failed at D:\Ribasim\core\test\allocation_test.jl:334
  Expression: ≈(storage[stage_1], u_stage_1.(t[stage_1]), rtol = 0.0001)
   Evaluated: [500.0, 672.7963463605041] ≈ [500.0, 543.2] (rtol=0.0001)
```

**Comment by SouthEndMusic** [comment]
This might have to do with the fact that allocation uses the flows from the last `water_balance!` call, which depends (maybe among other things) on the used solver and might not correspond to the right state of the Model. To make this consistent, maybe always do a `water_balance!` call before allocation, which also would help with performing allocation at `t=0`.
Alternatively, we should first pick up the issue about using average flows and levels as input for allocation (https://github.com/Deltares/Ribasim/issues/1121, but this issue does not yet take basin levels into account).

**Comment by visr** [comment]
It actually ran allocation on the wrong times, that is fixed now.

File: core/src/validation.jl [hunk 1]
```diff
@@ -146,6 +146,17 @@ function sorted_table!(
     return table
 end
 
+function valid_config(config::Config)::Bool
```
Comment (SouthEndMusic) [inline core/src/validation.jl:149]: Now that you have this, can it be removed below?

https://github.com/Deltares/Ribasim/blob/e10f657bb56434881bfd3382ddce921b35cff9f5/core/src/model.jl#L50-L53

File: core/src/model.jl [hunk 2]
```diff
@@ -188,10 +191,51 @@ function SciMLBase.successful_retcode(model::Model)::Bool
...
+"""
+function SciMLBase.step!(model::Model, dt::Float64)::Model
+    (; config, integrator) = model
+    (; t) = integrator
+    # If we are at an allocation time, run allocation before the next physical
+    # layer timestep. This allows allocation over period (t, t + dt) to use variables
+    # set over BMI at time t before calling this function.
+    # Also, don't run allocation at t = 0 since there are no flows yet (#1389).
+    ntimes = t / config.allocation.timestep
+    if ntimes > 0 && round(ntimes) ≈ ntimes
```
Comment (SouthEndMusic) [inline core/src/model.jl:206]: I'd say the more common way to do this is

```julia
if t % config.allocation.timestep ≈ 0.0
   ... 
end
```

and the other check could just be `t > 0` if you actually need that.

**Review by SouthEndMusic — APPROVED** [review]
Looks good, left some small comments

File: core/src/model.jl (hunk 2 above)
Comment (SouthEndMusic) [reply core/src/model.jl:206]: Although the comparison tolerances work out differently

File: core/src/validation.jl (hunk 1 above)
Comment (visr) [reply core/src/validation.jl:149]: Right, I forgot, thanks.

File: core/src/model.jl (hunk 2 above)
Comment (visr) [reply core/src/model.jl]: That would only work if the t is right after, but not right before, right?
```
julia> prevfloat(3.0) % 0.5
0.49999999999999956
```

**Comment by SouthEndMusic** [comment]
> Oops

lol

**Pull request merged**